src/body.rs

```diff
@@ -459,15 +468,25 @@ impl AsyncRead for Body {
         cx: &mut Context<'_>,
         buf: &mut [u8],
     ) -> Poll<io::Result<usize>> {
-        Pin::new(&mut self.reader).poll_read(cx, buf)
+        let mut buf = match self.length {
+            None => buf,
+            Some(length) if length == self.bytes_read => return Poll::Ready(Ok(0)),
+            Some(length) => {
+                let max_len = (length - self.bytes_read).min(buf.len());
+                &mut buf[0..max_len]
```

Comment:
Will this panic if the body has more bytes than it specifies?

Reply:
nope, there's a test for that. we just won't read them from the underlying Read, which is the correct behavior for keepalive

Reply:
this line limits the length of the mutable slice that we pass into the inner read implementation, effectively telling it "give me a maximum of `max_len` bytes